### PR TITLE
[Feature] Siri Shortcuts

### PR DESCRIPTION
**Pull request by alexaubry**
This PR adds the Siri shortcuts feature. Intents are defined in the `MetroKit` framework and shared in the interested targets. We add an intent handler extension that connects to the data store to fetch and update the data. A new sheet has been added to allow the user to add and edit shortcuts.

Note: `IntentsUI` extension is out of scope for now.

The supported shortcuts are:
- Swipe MetroCard
- Check MetroCard balance

### CODE REVIEW

File: App/AppDelegate.swift [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import MetroKit
 import SwiftUI
+import Intents
```
Comment (alexaubry) [inline App/AppDelegate.swift]: Delete

File: App/Components/Views/SafeAreaSpacer.swift [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import SwiftUI
+
+/// A spacer whose min length is set to the safe area insets of the specified edge.
+struct SafeAreaSpacer: View {
```
Comment (alexaubry) [inline App/Components/Views/SafeAreaSpacer.swift]: No longer needed

File: MetroKit/Intents/IntentDonationCenter.swift [hunk 3]
```diff
@@ -0,0 +1,8 @@
+//
```
Comment (alexaubry) [inline MetroKit/Intents/IntentDonationCenter.swift]: Delete this file

File: MetroKit/Core Data/MetroDataStore.swift [hunk 4]
```diff
@@ -10,8 +10,7 @@ public protocol MetroCardDataStore {
     func publisher(for card: ObjectReference<MetroCard>) -> AnyPublisher<ObjectReference<MetroCard>, Never>
     
     /// Update the data on the specified card reference by using the given update descriptior.
-    func applyUpdates(_ update: [MetroCardUpdate],
-                      to cardReference: ObjectReference<MetroCard>) -> AnyPublisher<Void, Error>
+    func applyUpdates(_ update: [MetroCardUpdate], to cardReference: ObjectReference<MetroCard>) -> AnyPublisher<Void, Error>
```
Comment (alexaubry) [inline MetroKit/Core Data/MetroDataStore.swift]: Typo update**s**

File: MetroIntents/IntentHandler.swift [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+class IntentHandler: INExtension {
+    override func handler(for intent: INIntent) -> Any {
+        let dataStore = try! PersistentMetroCardDataStore(
+            persistentStore: .sharedContainer,
+            useCloudKit: true
+        )
+
+        if intent is MBYSwipeCardIntent {
+            return SwipeMetroCardIntentHandler(dataStore: dataStore)
+        }
```
Comment (alexaubry) [inline MetroIntents/IntentHandler.swift:16]: TODO: Handle `MBYCheckBalanceIntent`

File: Configs/MetroIntents.Info.plist [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+	<string>$(CURRENT_PROJECT_VERSION)</string>
+	<key>MBYAppGroupName</key>
+	<string>group.$(DOMAIN)</string>
+	<key>NSExtension</key>
+	<dict>
+		<key>NSExtensionAttributes</key>
+		<dict>
+			<key>IntentsRestrictedWhileLocked</key>
+			<array>
+				<string>MBYSwipeCardIntent</string>
```
Comment (alexaubry) [inline Configs/MetroIntents.Info.plist:32]: Add `MBYCheckBalanceIntent` to restricted list

File: Configs/Base.xcconfig [hunk 7]
```diff
@@ -1,7 +1,7 @@
 DOMAIN = dev.alexisonline.MetroBuddy;
 APP_GROUP_ID = group.$(DOMAIN);
-SHORT_VERSION_STRING = 1.0.0;
-CURRENT_PROJECT_VERSION = 4;
+SHORT_VERSION_STRING = 1.1.0;
```
Comment (alexaubry) [inline Configs/Base.xcconfig]: Make these changes on `main` branch?

File: Configs/App.Info.plist [hunk 8]
```diff
@@ -24,6 +24,13 @@
 	<true/>
 	<key>MBYAppGroupName</key>
 	<string>group.$(DOMAIN)</string>
+	<key>NSSiriUsageDescription</key>
+	<string>This lets you ask Siri to update your MetroCard balance.</string>
```
Comment (alexaubry) [inline Configs/App.Info.plist]: Find better copy

File: App/Screens/Card/MetroCardViewModel.swift [hunk 9]
```diff
@@ -193,6 +197,34 @@ class MetroCardViewModel: ObservableObject {
...
         return update(number)
     }
+
+    // MARK: - Intents
+
+    func makeShortcutListViewModel() -> ShortcutListViewModel {
+        return ShortcutListViewModel(voiceShortcutsCenter: INVoiceShortcutCenter.shared)
+    }
+
+    private static func donateSwipeInteraction(requestedBalance: Decimal, completion: Subscribers.Completion<Error>) {
```
Comment (alexaubry) [inline App/Screens/Card/MetroCardViewModel.swift]: Extract to separate object

File: App/Screens/Siri/Views/Components/AddVoiceShortcutView.swift [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+        init(dismiss: @escaping () -> Void) {
+            self.dismiss = dismiss
+        }
+
+        func addVoiceShortcutViewControllerDidCancel(_ controller: INUIAddVoiceShortcutViewController) {
+            dismiss()
+        }
+
+        func addVoiceShortcutViewController(_ controller: INUIAddVoiceShortcutViewController, didFinishWith voiceShortcut: INVoiceShortcut?, error: Error?) {
+            #warning("TODO Alexis: Handle error")
```
Comment (alexaubry) [inline App/Screens/Siri/Views/Components/AddVoiceShortcutView.swift]: Handle error

File: App/Screens/Siri/Views/Components/EditVoiceShortcutView.swift [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+        init(dismiss: @escaping () -> Void) {
+            self.dismiss = dismiss
+        }
+
+        func editVoiceShortcutViewControllerDidCancel(_ controller: INUIEditVoiceShortcutViewController) {
+            dismiss()
+        }
+
+        func editVoiceShortcutViewController(_ controller: INUIEditVoiceShortcutViewController, didUpdate voiceShortcut: INVoiceShortcut?, error: Error?) {
+            #warning("TODO Alexis: Handle error")
```
Comment (alexaubry) [inline App/Screens/Siri/Views/Components/EditVoiceShortcutView.swift]: Handle error

File: App/Components/Sheets/ModalSheet.swift [hunk 12]
```diff
@@ -1,8 +1,25 @@
 import SwiftUI
 
+struct ModalBackdrop: View {
```
Comment (alexaubry) [inline App/Components/Sheets/ModalSheet.swift]: Refactor to allow animations

File: App/Components/Sheets/ModalDrawer.swift [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+private struct ModalDrawerContainer<Source: View, Sheet: View>: View {
+    let source: Source
+    let content: Sheet
+    @Binding var isPresented: Bool
+
+    var body: some View {
+        ZStack(alignment: .bottom) {
+            source
+                .zIndex(0)
+                .accessibility(hidden: isPresented)
```
Comment (a2) [inline App/Components/Sheets/ModalDrawer.swift:13]: Is there a SwiftUI equivalent for [`accessibilityViewIsModal`](https://developer.apple.com/documentation/objectivec/nsobject/1615089-accessibilityviewismodal)?

File: Configs/Base.xcconfig [hunk 14]
```diff
@@ -1,7 +1,7 @@
 DOMAIN = dev.alexisonline.MetroBuddy;
 APP_GROUP_ID = group.$(DOMAIN);
-SHORT_VERSION_STRING = 1.1.0;
-CURRENT_PROJECT_VERSION = 5;
+SHORT_VERSION_STRING = 1.0.0;
```
Comment (a2) [inline Configs/Base.xcconfig]: Was this intentional?

File: MetroKit/Core Data/MetroCardDataStoreError.swift [hunk 15]
```diff
@@ -31,7 +31,7 @@ public enum MetroCardDataStoreError: UserFacingError {
 
         case .cannotSave(let error):
             let format = NSLocalizedString(
-                "You changes could not be saved, as an error occured while saving. The error code is %@ %l. We suggest trying to force-quit the app and reopening it.",
+                "You changes could not be saved due to an error. The error code is %1$@ %2$ld. We suggest trying to force-quit the app and reopening it.",
```
Comment (a2) [inline MetroKit/Core Data/MetroCardDataStoreError.swift]:
```suggestion
                "Your changes could not be saved due to an error. The error code is %1$@ %2$ld. We suggest trying to force-quit the app and reopening it.",
```

File: MetroKit/Core Data/PersistentHistoryService.swift [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+        let fetchHistoryRequest = NSPersistentHistoryChangeRequest.fetchHistory(after: lastMerge)
+
+        do {
+            let historyResult = try context.execute(fetchHistoryRequest) as! NSPersistentHistoryResult
+            let history = historyResult.result as! [NSPersistentHistoryTransaction]
+
+            for transaction in history.reversed() {
+                context.stalenessInterval = 0
+                context.mergeChanges(fromContextDidSave: transaction.objectIDNotification())
+                context.stalenessInterval = -1
```
Comment (a2) [inline MetroKit/Core Data/PersistentHistoryService.swift]: Maybe store the old value / do you want to change this in each loop?

```swift
let previousStalenessInterval = context.stalenessInterval
for transaction in history.reversed() {
    context.mergeChanges(fromContextDidSave: transaction.objectIDNotification())
}
context.stalenessInterval = previousStalenessInterval
```

File: App/Components/Sheets/ModalDrawer.swift (hunk 13 above)
Comment (alexaubry) [reply App/Components/Sheets/ModalDrawer.swift:13]: Seems like you need to add `isModal` as an accessibility trait.

Comment (alexaubry) [reply App/Components/Sheets/ModalDrawer.swift:13]: Actually, it seems like that's not working for some reason? Will leave it like this.

**Pull request merged**